File: petclinic-test/src/genseq/steps.ts
// The sentence the test was on when a call went out.
//
// A diagram drawn from traces alone answers "what did the system do". It cannot answer
// "which line of the test made it do that" — the traces carry HTTP routes and SQL, and
// nothing in them knows the scenario was on `When I search owners for ""`. So the test
// stamps its own sentences as it walks through them, and the renderer folds those marks
// back in between the traces: one self-call on the driving lifeline, above the arrows
// that sentence caused.
//
// A timestamp, not a span: the browser and the backend are already traced by two
// different exporters, and threading a third context through Playwright's Node process
// into the page would buy nothing here. Both clocks are the same machine's, the marks
// and the traces are minutes apart at worst, and a step boundary only has to land
// between two HTTP calls — a resolution of milliseconds, against calls that are tens of
// milliseconds apart.
//
// (The @SpringBootTest half of this does use a span, because there it is free: the test
// and the code under test share one JVM and one OTel context. See
// petclinic-backend/src/test/java/victor/training/petclinic/genseq/.)

/** One sentence of a scenario, stamped when it started. */
export interface StepMark {
  label: string;
  atMs: number;
}

/** `openOwnerDetailPage` → `open owner detail page`. */
export function sentenceOf(name: string): string {
  return name
    .replace(/([a-z0-9])([A-Z])/g, '$1 $2')
    // `anOwnerWithAtLeastOnePetExistsHTTP` — an acronym run keeps its last letter for the
    // word that follows it, so `HTTPRequest` reads `HTTP request` and not `HTTPR equest`.
    .replace(/([A-Z]+)([A-Z][a-z])/g, '$1 $2')
    .replace(/[_-]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .toLowerCase();
}

/**
 * The marks of the scenario currently running.
 *
 * One recorder per process, deliberately: Playwright gives each worker its own process
 * and runs one test at a time in it, and Cucumber runs a whole file serially. Two
 * scenarios can therefore never be writing here at once — which is the same reasoning
 * that lets the trace window be a single start/end pair per scenario.
 */
export class StepRecorder {
  private marks: StepMark[] = [];

  constructor(private readonly now: () => number = Date.now) {}

  /** Begin a scenario: whatever the last one left behind is not this one's. */
  start(): void {
    this.marks = [];
  }

  mark(label: string): void {
    const text = label.trim();
    if (!text) return;
    // A sentence repeated back-to-back is one step, not two: the diagram would draw the
    // same narration twice with nothing between the copies to explain the repetition.
    if (this.marks[this.marks.length - 1]?.label === text) return;
    this.marks.push({label: text, atMs: this.now()});
  }

  /** Hand over what this scenario recorded, and forget it. */
  take(): StepMark[] {
    const taken = this.marks;
    this.marks = [];
    return taken;
  }
}

/** What the narrated DSL and the Cucumber hooks write into. */
export const steps = new StepRecorder();

/**
 * The same sentences, each announcing itself before it runs.
 *
 * `add-visit.spec.ts` imports its DSL through this, so the spec body stays exactly the
 * list of sentences it was — no `test.step()` wrapper round every line, no second name
 * for anything. The sentence on the diagram is the function's own name, so it cannot
 * drift from the code: rename the function and the narration renames itself.
 */
export function narrate<T extends object>(sentences: T, recorder: StepRecorder = steps): T {
  const narrated: Record<string, unknown> = {};
  for (const [name, value] of Object.entries(sentences)) {
    narrated[name] = typeof value !== 'function' ? value : (...args: unknown[]) => {
      // Marked before the call, not after: the mark is when the sentence *started*, and
      // every call it makes happens after that instant. Stamping on the way out would
      // put the sentence below its own arrows.
      recorder.mark(sentenceOf(name));
      return (value as (...a: unknown[]) => unknown)(...args);
    };
  }
  return narrated as T;
}

/** The sentence a trace belongs to: the last one that had started when the trace opened. */
export function stepAt(marks: readonly StepMark[], atMs: number): StepMark | undefined {
  let found: StepMark | undefined;
  for (const mark of marks) {
    if (mark.atMs > atMs) break;
    found = mark;
  }
  return found;
}

File: petclinic-backend/src/test/java/victor/training/petclinic/genseq/Steps.java
package victor.training.petclinic.genseq;

import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.context.Scope;

/**
 * The sentences of a @SpringBootTest, so the generated diagram can say which one caused which call.
 * <p>
 * A mark, not a block:
 *
 * <pre>
 *     given("an owner with one pet");
 *     int ownerId = ownerRepository.save(anOwnerWithAPet()).getId();
 *
 *     when("the owner is fetched");
 *     mockMvc.perform(get("/api/owners/{id}", ownerId));
 * </pre>
 *
 * A <code>given(String, Runnable)</code> would have scoped itself properly, but it forces every
 * value the next sentence needs out of the test's own locals and into a field or a holder — and a
 * test that has to be restructured to be drawn is a test nobody will draw. Each mark simply closes
 * the span the previous one opened, which is the same shape the TypeScript side uses (a timestamp
 * per sentence, see petclinic-test/src/genseq/steps.ts) — only here the mark can be a real span,
 * because the test and the code under test share one JVM and therefore one OTel context.
 * <p>
 * Everything the sentence goes on to do — the MockMvc call, the repository save, the SQL underneath
 * it — lands inside that span, and that parentage is what the renderer draws as an arrow leaving
 * the Test lifeline.
 */
public final class Steps {

    /**
     * The lifeline this span belongs on. Nothing else in the trace can tell the test apart from the
     * code it drives: both run in one JVM under one <code>service.name</code>, so without this the
     * whole diagram collapses onto a single participant. Read by petclinic-test's trace-to-puml.ts.
     */
    static final AttributeKey<String> PARTICIPANT = AttributeKey.stringKey("genseq.participant");

    static final String TEST_PARTICIPANT = "Test";

    /** The step currently open on this thread — JUnit runs a test's before/body/after on one. */
    private static final ThreadLocal<Open> current = new ThreadLocal<>();

    private record Open(Span span, Scope scope) {
    }

    private Steps() {
    }

    public static void given(String sentence) {
        announce("given " + sentence);
    }

    public static void when(String sentence) {
        announce("when " + sentence);
    }

    public static void then(String sentence) {
        announce("then " + sentence);
    }

    /** Continues whichever of the three came before it, exactly as Gherkin's `And` does. */
    public static void and(String sentence) {
        announce("and " + sentence);
    }

    private static void announce(String sentence) {
        close();
        Span span = GlobalOpenTelemetry.getTracer("petclinic-genseq")
                .spanBuilder(sentence)
                .setAttribute(PARTICIPANT, TEST_PARTICIPANT)
                .startSpan();
        // The scope is deliberately held open across the statements that follow rather than
        // closed by a try-with-resources: the sentence lasts until the next sentence starts.
        current.set(new Open(span, span.makeCurrent()));
    }

    /** Closes the last sentence of a test. Called by {@link SequenceTraceExtension} on the way out. */
    static void close() {
        Open open = current.get();
        if (open == null) {
            return;
        }
        current.remove();
        // Scopes must be closed on the thread that opened them, innermost first — which is exactly
        // what this is, since a new sentence never opens before the previous one has closed.
        open.scope().close();
        open.span().end();
    }
}
